rust/kernel/hw_random.rs

```diff
@@ -0,0 +1,200 @@
+/// be set to null pointers.
+pub const USE_NONE: ToUse = ToUse {
+    init: false,
+    cleanup: false,
+};
+
+/// Defines the [`HwrngOperations::TO_USE`] field based on a list of fields to be populated.
+#[macro_export]
+macro_rules! declare_hwrng_operations {
+    () => {
+        const TO_USE: $crate::hw_random::ToUse = $crate::hw_random::USE_NONE;
+    };
+    ($($i:ident),+) => {
+        const TO_USE: kernel::hw_random::ToUse =
+            $crate::hw_random::ToUse {
+                $($i: true),+ ,
+                ..$crate::hw_random::USE_NONE
+            };
+    };
+}
+
+unsafe extern "C" fn init_callback<T: HwrngOperations>(
+    rng: *mut bindings::hwrng,
+) -> c_types::c_int {
+    from_kernel_result! {
+        // SAFETY: `priv` private data field was initialized during creation of
+        // the `bindings::hwrng` in `Self::init_hwrng` function. This callback
+        // is only called once `new_pinned` suceeded previously which guarantees safety.
+        let data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+        T::init(data)?;
+        Ok(0)
+    }
+}
+
+unsafe extern "C" fn cleanup_callback<T: HwrngOperations>(rng: *mut bindings::hwrng) {
+    // SAFETY: `priv` private data field was initialized during creation of
+    // the `bindings::hwrng` in `Self::init_hwrng` function. This callback
+    // is only called once `new_pinned` suceeded previously which guarantees safety.
+    let data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+    T::cleanup(data);
+}
+
+unsafe extern "C" fn read_callback<T: HwrngOperations>(
+    rng: *mut bindings::hwrng,
+    data: *mut c_types::c_void,
+    max: usize,
+    wait: bindings::bool_,
+) -> c_types::c_int {
+    from_kernel_result! {
+        // SAFETY: `priv` private data field was initialized during creation of
+        // the `bindings::hwrng` in `Self::init_hwrng` function. This callback
+        // is only called once `new_pinned` suceeded previously which guarantees safety.
+        let drv_data = unsafe { T::Data::borrow((*rng).priv_ as *const core::ffi::c_void) };
+
+        // SAFETY: Slice is created from `data` and `max` arguments that are C api buffer
+        // and its size in bytes which are safe for this conversion.
+        let buffer = unsafe { from_raw_parts_mut(data as *mut c_types::c_char, max) };
+        let ret = T::read(drv_data, buffer, wait)?;
+        Ok(ret)
+    }
+}
+
+// SAFETY: `Registration` does not expose any of its state across threads.
+unsafe impl<T: HwrngOperations> Sync for Registration<T> {}
+
+impl<T: HwrngOperations> Drop for Registration<T> {
+    /// Removes the registration from the kernel if it has completed successfully before.
+    fn drop(&mut self) {
+        // SAFETY: The instance of Registration<T> is returned from `new_pinned`
+        // function after being initialized and registered.
+        unsafe { bindings::hwrng_unregister(&mut *self.hwrng.get()) };
```

Comment:
What prevents you from calling this even when `hwrng_register` hasn't been called or failed?

Also, `self.hwrng.get()` should suffice as the argument.

Reply:
I see the issue, thanks.